File: frontend/src/api/userProfile.ts
export type ProfilePersonal = {
  user_id: number;
  name: string;
  login: string;
  email: string | null;
  phone: string | null;
  id_doc: string | null;
  is_juridical: number;
  entity_label: string;
  user_status: number | null;
  status_label: string;
  station_name: string | null;
  auth_page: string | null;
};

export type ProfileOnline = {
  is_online: boolean;
  last_session_end: string | null;
  last_session_end_label: string | null;
};

export type ProfileTariff = {
  state: "active" | "inactive" | "frozen" | "planned_freeze";
  tariff_name: string;
  real_type: string | null;
  is_active: boolean;
  rate_up: string;
  rate_down: string;
  speed_unlimited: boolean;
  remain_traffic_mb: number;
  full_packet_mb: number;
  jur_main_packet_mb: number | null;
  jur_dop_packet_mb: number | null;
  overrun_mb: number | null;
  traffic_renew_count: number | null;
  msk_reset: string | null;
  local_reset: string | null;
  last_traffic_reset_label: string | null;
  valid_date_label: string | null;
  disconnect_at_label: string | null;
  planned_freeze_at: string | null;
  frozen_at: string | null;
  unfreeze_at: string | null;
  frozen_remaining_label: string | null;
  freeze_reason: string | null;
  can_freeze: boolean;
  can_unfreeze: boolean;
  can_cancel_planned_freeze: boolean;
  can_remove_ended_tariff: boolean;
  can_disconnect_sessions: boolean;
};

export type ProfileTicket = {
  id: number;
  title: string;
  category: string | null;
  category_theme: string | null;
  date_of_create: string;
  date_of_close: string | null;
  assigned_to_role: string | null;
  support_line: number;
  support_line_label: string;
  status: string;
  status_label: string;
};

export type ProfileHealthCheck = { items: string[] };

export type UserProfileResponse = {
  personal: ProfilePersonal;
  online: ProfileOnline;
  open_sessions_count: number;
  balance: number;
  tariff: ProfileTariff | null;
  netflow_note: string | null;
  netflow_tariff: string | null;
  health_check: ProfileHealthCheck;
  tickets: ProfileTicketListResponse;
  disconnect_sessions_remaining: number;
  disconnect_sessions_limit: number;
  disconnect_sessions_window_minutes: number;
};

export type TariffBlockResponse = {
  ok: boolean;
  message: string;
  tariff: ProfileTariff | null;
  netflow_note: string | null;
  netflow_tariff: string | null;
  health_check: ProfileHealthCheck;
};

export type ProfileTicketListResponse = {
  total: number;
  page: number;
  per_page: number;
  items: ProfileTicket[];
};

function formatApiDetail(detail: unknown, status: number): string {
  if (typeof detail === "string") return detail;
  if (Array.isArray(detail)) {
    const msgs = detail.map((item) => {
      if (item && typeof item === "object" && "msg" in item) {
        return String((item as { msg: string }).msg);
      }
      return String(item);
    });
    if (msgs.length) return msgs.join("; ");
  }
  return `HTTP ${status}`;
}

async function api<T>(url: string, init?: RequestInit): Promise<T> {
  const res = await fetch(url, { credentials: "include", ...init });
  const data = (await res.json().catch(() => ({}))) as { detail?: unknown } & Partial<T>;
  if (!res.ok) {
    throw new Error(formatApiDetail(data.detail, res.status));
  }
  return data as T;
}

export function fetchUserProfile(
  userId: number,
  ticketsPage = 1,
  ticketsPerPage = 10,
  includeTickets = true,
): Promise<UserProfileResponse> {
  const q = new URLSearchParams({
    tickets_page: String(ticketsPage),
    tickets_per_page: includeTickets ? String(ticketsPerPage) : "0",
    include_tickets: includeTickets ? "true" : "false",
  });
  return api(`/api/v1/helpdesk/users/${userId}/profile?${q}`);
}

export function fetchUserProfileTickets(
  userId: number,
  page = 1,
  perPage = 10,
): Promise<ProfileTicketListResponse> {
  const q = new URLSearchParams({ page: String(page), per_page: String(perPage) });
  return api(`/api/v1/helpdesk/users/${userId}/profile/tickets?${q}`);
}

export function postUnarchive(userId: number) {
  return api<{ message: string }>(`/api/v1/helpdesk/users/${userId}/unarchive`, { method: "POST" });
}

export function postUnfreeze(userId: number) {
  return api<{ message: string }>(`/api/v1/helpdesk/users/${userId}/unfreeze`, { method: "POST" });
}

export function deleteFreezePlan(userId: number) {
  return api<{ message: string }>(`/api/v1/helpdesk/users/${userId}/freeze-plan`, { method: "DELETE" });
}

export function postFreeze(userId: number, body: { date_freeze?: string | null; date_unfreeze?: string | null }) {
  return api<{ message: string }>(`/api/v1/helpdesk/users/${userId}/freeze`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
}

export function postDisconnect(userId: number) {
  return api<{ message: string }>(`/api/v1/helpdesk/users/${userId}/disconnect-sessions`, { method: "POST" });
}

export function postRemoveEndedTariff(userId: number) {
  return api<TariffBlockResponse>(`/api/v1/helpdesk/users/${userId}/tariff/remove-ended`, {
    method: "POST",
  });
}

export type FastCheckStatus = "pass" | "fail" | "warn" | "skip";

export type FastCheckStep = {
  test_code: string;
  variant: number;
  check_label: string;
  status: FastCheckStatus;
  detail?: string | null;
  actions_html?: string | null;
  stop_chain: boolean;
};

export type ManagerContact = {
  full_name?: string | null;
  phones: string[];
  emails: string[];
};

export type FastCheckResponse = {
  steps: FastCheckStep[];
  stopped_at?: string | null;
  manager_contacts: ManagerContact[];
};

export function postFastCheck(userId: number): Promise<FastCheckResponse> {
  return api<FastCheckResponse>(`/api/v1/helpdesk/users/${userId}/fast-check`, { method: "POST" });
}

export type PaymentHistoryItem = {
  msk_date: string;
  msk_date_label: string;
  state: string;
  state_label: string;
  payment_type: string;
  type_label: string;
  amount: number;
};

export type PaymentHistoryListResponse = {
  total: number;
  page: number;
  per_page: number;
  items: PaymentHistoryItem[];
};

export function fetchUserPayments(
  userId: number,
  page = 1,
  perPage = 10,
): Promise<PaymentHistoryListResponse> {
  const q = new URLSearchParams({
    page: String(page),
    per_page: String(perPage),
  });
  return api<PaymentHistoryListResponse>(
    `/api/v1/helpdesk/users/${userId}/payments?${q}`,
  );
}

export type TariffHistoryItem = {
  activated_at: string;
  activated_at_label: string;
  row_kind: "tariff" | "dop";
  type_label: string;
  type_hint: string | null;
  active_tariff: boolean;
  deactivation_at_label: string | null;
  price: number | null;
  price_label: string;
};

export type TariffHistoryListResponse = {
  total: number;
  page: number;
  per_page: number;
  items: TariffHistoryItem[];
};

export function fetchUserTariffHistory(
  userId: number,
  page = 1,
  perPage = 10,
): Promise<TariffHistoryListResponse> {
  const q = new URLSearchParams({
    page: String(page),
    per_page: String(perPage),
  });
  return api<TariffHistoryListResponse>(
    `/api/v1/helpdesk/users/${userId}/tariff-history?${q}`,
  );
}

export type PasswordResetState = {
  has_ppp_sessions: boolean;
  active_code: string | null;
  expires_at: string | null;
  can_generate: boolean;
  code_id: number | null;
};

export type PasswordResetGenerateResult = {
  code: string;
  expires_at: string;
  code_id: number;
  message: string;
};

export type PasswordResetPoll = {
  code_used: boolean;
  code_expired: boolean;
  active_code: string | null;
  expires_at: string | null;
  code_id: number | null;
};

export function fetchPasswordResetState(userId: number): Promise<PasswordResetState> {
  return api(`/api/v1/helpdesk/users/${userId}/password-reset`);
}

export function fetchPasswordResetPoll(userId: number, codeId: number): Promise<PasswordResetPoll> {
  return api(`/api/v1/helpdesk/users/${userId}/password-reset/poll?code_id=${codeId}`);
}

export function generatePasswordResetCode(userId: number): Promise<PasswordResetGenerateResult> {
  return api(`/api/v1/helpdesk/users/${userId}/password-reset/generate`, { method: "POST" });
}

export type TrafficDetailSendRequest = {
  date_from: string;
  date_to: string;
};

export type TrafficDetailSendResponse = {
  message: string;
  email: string;
};

export function postTrafficDetailSend(
  userId: number,
  body: TrafficDetailSendRequest,
): Promise<TrafficDetailSendResponse> {
  return api<TrafficDetailSendResponse>(`/api/v1/helpdesk/users/${userId}/traffic-detail/send`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
}

File: app/api/v1/routers/helpdesk/user_profile.py
from __future__ import annotations

from fastapi import APIRouter, Depends, Query, Request
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.v1.routers.helpdesk.deps import require_tracker_user
from app.api.v1.routers.helpdesk.password_reset_schemas import (
    PasswordResetGenerateResponse,
    PasswordResetPollResponse,
    PasswordResetStateResponse,
)
from app.api.v1.routers.helpdesk.fast_check_schemas import FastCheckResponse
from app.api.v1.routers.helpdesk.user_profile_schemas import (
    ActionMessage,
    FreezeRequest,
    PaymentHistoryListResponse,
    TariffBlockResponse,
    TariffHistoryListResponse,
    ProfileTicketListResponse,
    UserProfileResponse,
)
from app.api.v1.routers.helpdesk import fast_check_service as fc_svc
from app.api.v1.routers.helpdesk import password_reset_service as pwd_svc
from app.api.v1.routers.helpdesk import user_profile_service as svc
from app.database import get_db

router = APIRouter(prefix="/v1/helpdesk/users", tags=["Helpdesk — абонент"])


@router.get("/{user_id}/payments", response_model=PaymentHistoryListResponse)
async def list_user_payments(
    user_id: int,
    page: int = Query(1, ge=1),
    per_page: int = Query(10, ge=1, le=50),
    _user: dict = Depends(require_tracker_user),
    db: AsyncSession = Depends(get_db),
) -> PaymentHistoryListResponse:
    return await svc.load_user_payments_page(db, user_id, page=page, per_page=per_page)


@router.get("/{user_id}/tariff-history", response_model=TariffHistoryListResponse)
async def list_user_tariff_history(
    user_id: int,
    page: int = Query(1, ge=1),
    per_page: int = Query(10, ge=1, le=50),
    _user: dict = Depends(require_tracker_user),
    db: AsyncSession = Depends(get_db),
) -> TariffHistoryListResponse:
    return await svc.load_user_tariff_history_page(db, user_id, page=page, per_page=per_page)


@router.get("/{user_id}/profile/tickets", response_model=ProfileTicketListResponse)
async def list_user_profile_tickets(
    user_id: int,
    page: int = Query(1, ge=1),
    per_page: int = Query(10, ge=1, le=50),
    _user: dict = Depends(require_tracker_user),
    db: AsyncSession = Depends(get_db),
) -> ProfileTicketListResponse:
    return await svc.load_user_tickets_page(db, user_id, page=page, per_page=per_page)


@router.get("/{user_id}/profile", response_model=UserProfileResponse)
async def get_profile(
    user_id: int,
    tickets_page: int = Query(1, ge=1),
    tickets_per_page: int = Query(10, ge=0, le=50),
    include_tickets: bool = Query(False),
    _user: dict = Depends(require_tracker_user),
    db: AsyncSession = Depends(get_db),
) -> UserProfileResponse:
    return await svc.get_user_profile(
        db,
        user_id,
        tickets_page,
        tickets_per_page,
        include_tickets=include_tickets and tickets_per_page > 0,
    )


@router.post("/{user_id}/unarchive", response_model=ActionMessage)
async def unarchive(
    user_id: int,
    _user: dict = Depends(require_tracker_user),
    db: AsyncSession = Depends(get_db),
) -> ActionMessage:
    return await svc.unarchive_user(db, user_id)


@router.post("/{user_id}/unfreeze", response_model=ActionMessage)
async def unfreeze(
    user_id: int,
    request: Request,
    operator: dict = Depends(require_tracker_user),
    db: AsyncSession = Depends(get_db),
) -> ActionMessage:
    return await svc.unfreeze_tariff(db, user_id, operator, request)


@router.delete("/{user_id}/freeze-plan", response_model=ActionMessage)
async def cancel_freeze_plan(
    user_id: int,
    request: Request,
    operator: dict = Depends(require_tracker_user),
    db: AsyncSession = Depends(get_db),
) -> ActionMessage:
    return await svc.delete_planned_freeze(db, user_id, operator, request)


@router.post("/{user_id}/tariff/remove-ended", response_model=TariffBlockResponse)
async def remove_ended_tariff(
    user_id: int,
    request: Request,
    operator: dict = Depends(require_tracker_user),
    db: AsyncSession = Depends(get_db),
) -> TariffBlockResponse:
    return await svc.remove_ended_tariff(db, user_id, operator, request)


@router.post("/{user_id}/freeze", response_model=ActionMessage)
async def freeze(
    user_id: int,
    body: FreezeRequest,
    request: Request,
    operator: dict = Depends(require_tracker_user),
    db: AsyncSession = Depends(get_db),
) -> ActionMessage:
    return await svc.apply_freeze(
        db, user_id, body.date_freeze, body.date_unfreeze, operator, request
    )


@router.get("/{user_id}/password-reset", response_model=PasswordResetStateResponse)
async def password_reset_state(
    user_id: int,
    request: Request,
    operator: dict = Depends(require_tracker_user),
    db: AsyncSession = Depends(get_db),
) -> PasswordResetStateResponse:
    return await pwd_svc.get_password_reset_state(db, user_id, operator, request)


@router.post("/{user_id}/password-reset/generate", response_model=PasswordResetGenerateResponse)
async def password_reset_generate(
    user_id: int,
    request: Request,
    operator: dict = Depends(require_tracker_user),
    db: AsyncSession = Depends(get_db),
) -> PasswordResetGenerateResponse:
    return await pwd_svc.generate_password_reset_code(db, user_id, operator, request)


@router.get("/{user_id}/password-reset/poll", response_model=PasswordResetPollResponse)
async def password_reset_poll(
    user_id: int,
    code_id: int | None = None,
    _operator: dict = Depends(require_tracker_user),
    db: AsyncSession = Depends(get_db),
) -> PasswordResetPollResponse:
    return await pwd_svc.poll_password_reset_status(db, user_id, code_id)


@router.post("/{user_id}/fast-check", response_model=FastCheckResponse)
async def fast_check(
    user_id: int,
    _user: dict = Depends(require_tracker_user),
    db: AsyncSession = Depends(get_db),
) -> FastCheckResponse:
    return await fc_svc.run_fast_check(db, user_id)


@router.post("/{user_id}/disconnect-sessions", response_model=ActionMessage)
async def disconnect_sessions(
    user_id: int,
    _user: dict = Depends(require_tracker_user),
    db: AsyncSession = Depends(get_db),
) -> ActionMessage:
    from app.api.v1.routers.users.dao import UsersDAO

    u = await UsersDAO.find_one_or_none(db, id=user_id)
    if not u:
        from fastapi import HTTPException

        raise HTTPException(status_code=404, detail="Абонент не найден")
    login = (u.get("login") or "").strip()
    return await svc.force_disconnect(db, user_id, login)
